pygmt/src/tilemap.py

```diff
@@ -0,0 +1,145 @@
+        source.
+
+    source : xyzservices.TileProvider or str
+        Optional. The tile source: web tile provider or path to a local file.
+        Provide either:
+
+        - A web tile provider in the form of a
+          :class:`xyzservices.TileProvider` object. See
+          :doc:`Contextily providers <contextily:providers_deepdive>` for a
+          list of tile providers [Default is
+          ``xyzservices.providers.Stamen.Terrain``, i.e. Stamen Terrain web
+          tiles].
+        - A web tile provider in the form of a URL. The placeholders for the
+          XYZ in the URL need to be {{x}}, {{y}}, {{z}}, respectively. E.g.
+          ``https://{{s}}.tile.openstreetmap.org/{{z}}/{{x}}/{{y}}.png``.
+        - A local file path. The file is read with
+          :doc:`rasterio <rasterio:index>` and all bands are loaded into the
+          basemap. See
+          :doc:`contextily:working_with_local_files`.
+
+        IMPORTANT: Tiles are assumed to be in the Spherical Mercator projection
+        (EPSG:3857).
+
+    lonlat : bool
+        Optional. If ``False``, coordinates in ``region`` are assumed to be
+        Spherical Mercator as opposed to longitude/latitude [Default is
+        ``True``].
+
+    wait : int
+        Optional. If the tile API is rate-limited, the number of seconds to
+        wait between a failed request and the next try [Default is ``0``].
+
+    max_retries : int
+        Optional. Total number of rejected requests allowed before contextily
+        will stop trying to fetch more tiles from a rate-limited API [Default
+        is ``2``].
+
+    kwargs : dict
+        Extra keyword arguments to pass to :meth:`pygmt.Figure.grdimage`.
+
+    Raises
+    ------
+    ModuleNotFoundError
+        If ``rioxarray`` is not installed. Follow
+        :doc:`install instructions for rioxarray <rioxarray:installation>`,
+        (e.g. via ``pip install rioxarray``) before using this function.
+    """
+    from pygmt.datasets import load_tile_map  # pylint: disable=import-outside-toplevel
+
+    kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
+
+    if rioxarray is None:
+        raise ModuleNotFoundError(
+            "Package `rioxarray` is required to be installed to use this function. "
+            "Please use `pip install rioxarray` or "
+            "`conda install -c conda-forge rioxarray` "
+            "to install the package."
+        )
+
+    with GMTTempFile(suffix=".tif") as tmpfile:
+        raster = load_tile_map(
+            region=region,
+            zoom=zoom,
+            source=source,
+            lonlat=lonlat,
+            wait=wait,
+            max_retries=max_retries,
+        )
+        raster.rio.to_raster(raster_path=tmpfile.name)
+
+        with Session() as lib:
+            lib.call_module(
+                module="grdimage", args=build_arg_string(kwargs, infile=tmpfile.name)
+            )
```

Comment:
Note that this is not the full implementation of PyGMT's `fig.grdimage`, which is actually a bit more complicated to support shading (`-I`) with an `xarray.DataArray` grid

https://github.com/GenericMappingTools/pygmt/blob/8f31706fdfdc5d3cf1b15e77fc64db304031b53e/pygmt/src/grdimage.py#L179-L192

Question is: should we support shading of the XYZ tiles with `xarray.DataArray` grids? Or is there a way to use the `grdimage` implementation above directly without code duplication?

Reply:
Currently I would say no to shading.